packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1228 @@
+/// receives the configuration through [setInitialRoutePath] or
+/// [setNewRoutePath] to configure itself and builds the latest navigating
+/// widget upon asked.
+///
+/// When implementing subclass, consider defining a listenable app
+/// state to be used for building the navigating widget. The router delegate
+/// should update the app state accordingly and notify the listener know the app
+/// state has changed when it receive route related engine intents (e.g.
+/// [setNewRoutePath], [setInitialRoutePath], or [popRoute]).
+///
+/// All subclass must implement [setNewRoutePath], [popRoute], and [build].
+///
+/// See also:
+///
+///  * [RouteInformationParser], which is responsible for parsing the route
+///    information to a configuration before passing in to router delegate.
+///  * [Router], which is the widget that wires all the delegates together to
+///    provide a fully functional routing solution.
+abstract class RouterDelegate<T> extends Listenable {
+  /// Called by the [Router] at startup with the structure that the
+  /// [RouteInformationParser] obtained from parsing the initial route.
+  ///
+  /// This should configure the [RouterDelegate] so that when [build] is
+  /// invoked, it will create a widget tree that matches the initial route.
+  ///
+  /// By default, this method forwards the [configuration] to [setNewRoutePath].
+  ///
+  /// Consider using a [SynchronousFuture] if the result can be computed
+  /// synchronously, so that the [Router] does not need to wait for the next
+  /// microtask to schedule a build.
+  Future<void> setInitialRoutePath(T configuration) {
+    return setNewRoutePath(configuration);
+  }
+
+  /// Called by the [Router] when the [Router.routeInformationProvider] reports that a
+  /// new route has been pushed to the application by the operating system.
+  ///
+  /// Consider using a [SynchronousFuture] if the result can be computed
+  /// synchronously, so that the [Router] does not need to wait for the next
+  /// microtask to schedule a build.
+  Future<void> setNewRoutePath(T configuration);
+
+  /// Called by the [Router] when the [Router.backButtonDispatcher] reports that
+  /// the operating system is requesting that the current route be popped.
+  ///
+  /// The method should return a boolean [Future] to indicate whether this
+  /// delegate handles the request. Returning false will cause the entire app
+  /// to be popped.
+  ///
+  /// If the app state has changed as a result the pop, make sure to call the
+  /// [notifyListeners] so that the listener, typically the [Router], can
+  /// respond accordingly.
+  ///
+  /// Consider using a [SynchronousFuture] if the result can be computed
+  /// synchronously, so that the [Router] does not need to wait for the next
+  /// microtask to schedule a build.
+  Future<bool> popRoute();
+
+  /// Called by the [Router] when it detects a route information may have changed as a
```

Comment:
This should say something along the lines: When the configuration returned by this method is passed back to setInitialRoutePath or setNewRoutePath, the Router should configure the Navigator in `build` exactly like it did when the configuration was obtained.